sherpa/utils/random.py

```diff
@@ -0,0 +1,290 @@
+#
+#  Copyright (C) 2023
+#  Smithsonian Astrophysical Observatory
+#
+#
+#  This program is free software; you can redistribute it and/or modify
+#  it under the terms of the GNU General Public License as published by
+#  the Free Software Foundation; either version 3 of the License, or
+#  (at your option) any later version.
+#
+#  This program is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License along
+#  with this program; if not, write to the Free Software Foundation, Inc.,
+#  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+
+"""
+Basic routines for handling random numbers.
+
+A number of routines have been added to support writing code that
+allows both the legacy (pre 1.17) NumPy random API to be used
+(with rng set to None) or the new generator approach.
+
+"""
+
+import numpy as np
+
+# Copy of sherpa.utils.SherpaFloat
+#
+SherpaFloat = np.float_
+
+
+def poisson_noise(x, rng=None):
+    """Draw samples from a Poisson distribution.
+
+
+    Parameters
+    ----------
+    x : scalar or array
+       The expectation value for the distribution.
+    rng : np.random.Generator, np.random.RandomState, or None, optional
+        If set, the generator is used to create the random numbers. If
+        not set then the legacy numpy RandomState instance is used.
+
+    Returns
+    -------
+    out : scalar or array
+       A random realisation of the input array, drawn from
+       the Poisson distribution, as a `SherpaFloat`.
+
+    Notes
+    -----
+    The distribution is calculated by the NumPy poisson generator
+    and all values zero or less are replaced by zero.
+
+    Examples
+    --------
+
+    >>> poisson_noise([10, 20, 5])
+    array([ 13.,  21.,   6.])
+
+    >>> rng = np.random.default_rng(17389)
+    >>> poisson_noise([1, 2, 2, 3], rng=rng)
+    array([2., 2., 5., 3.])
+
+    """
+
+    if rng is None:
+        simulate = np.random.poisson
+    else:
+        simulate = rng.poisson
+
+    x = np.asarray(x)
+    if x.shape == ():
+        x = SherpaFloat(x)
+        if x <= 0.0:
+            x = 0.0
+        else:
+            x = x = simulate(x)
```

Comment:
Because it looked cool? Will fix (but fortunately, thanks to the way Python assignment works, it isn't a bug [note added for anyone reading along at home]).